docs/design/assembly-loading.md

```diff
@@ -0,0 +1,83 @@
+# ILLink Assembly Loading
+
+We would like to support loading and processing assemblies later in the pipeline, so that dynamically referenced assemblies can be correctly processed. A secondary goal is to avoid unnecessary processing of unused assemblies.
+
+## Assembly processing
+
+Assemblies loaded into the context usually need to have the following processing done before or during marking:
+- build the type hierarchy info
+- read descriptor XML and mark its new dependencies
+- read attribute XML and process it for additional/removed attributes
+- read substitutions XML and process it for constant methods
+- process the assembly IL for constant methods
+- remove unused branches after propagating constants (possibly from another assembly)
+
+## .NET5 behavior
+
+### Assembly references loaded up-front
+
+The linker processes a set of roots from command-line inputs (specific assemblies by name, or XML files that specify members to root). `LoadReferencesStep`, an early step in the linker pipeline, recursively loads all assemblies statically referenced by the root assemblies. The set of loaded assemblies is maintained in the `LinkContext`, and passed through each pipeline step in turn.
+
+Other early steps which load additional assemblies (for example, steps which process embedded XML, or scan for `DynamicDependencyAttribute`s referencing assemblies by name) are implemented as subclasses of `LoadReferencesStep`, calling into the same logic to recursively load referenced assemblies. However, these steps do not do recursive _processing_ of the referenced assemblies - so for example, an assembly referenced only by embedded XML will not have its own embedded XML processed.
+
+XML descriptors from the static reference closure are processed whether or not the embedding assembly is marked.
+
+### Dynamic assembly references not discovered
+
+Later steps do not load additional assemblies, with rare exceptions. Dynamic references (from reflection, or from special attributes) are only looked up in already-loaded assemblies, so the linker is unable to find assemblies only referenced dynamically, producing incorrect outputs which are missing the code required by such references. This problem is partially mitigated for `DynamicDependencyAttribute` by a step that pre-scans for assemblies referenced from instances of this attribute, but it does not recursively handle `DynamicDependencyAttribute` instances in these referenced assemblies.
+
+
+## .NET6 behavior
+
+### Assembly references loaded lazily
```

Comment:
I think it would be good to cover what will trigger the assembly reference loading

Reply:
Added some more description